[PHX-485] Support markdown in the Block API

micahhahn opened the pull request:
This PR adds support for markdown to the Block API in two places: block labels and block contents. 

For block labels we simply start interpreting the label string as markdown using existing the existing `markdown` feature of the Ballon API. 

For block contents I've added two new functions `bold` and `italic` that work only on the `Content` type.  This should give content-creation enough flexibility to handle all cases of markdown in question segments. 

## Context

Fixes [PHX-485](https://linear.app/noredink/issue/PHX-485/support-markdown-in-block-api)

## :framed_picture: What does this change look like?

<img width="888" alt="image" src="https://user-images.githubusercontent.com/12899207/229864483-59581e2f-7082-45ff-bb2d-f3fa3d0b1c08.png">

## Component completion checklist

- [x] I've gone through the relevant sections of the [Development Accessibility guide](https://paper.dropbox.com/doc/Accessibility-guide-4-Development--BiIVdijSaoijjOuhz3iTCJJ1Ag-rGoHpC91pFg3zTrYpvOCQ) with the changes I made to this component in mind
- [x] Changes are clearly documented
    - [x] Component docs include a changelog
    - [x] Any new exposed functions or properties have docs
- [x] Changes extend to the Component Catalog
    - [x] The Component Catalog is updated to use the newest version, if appropriate
    - [x] The Component Catalog example version number is updated, if appropriate
    - [x] Any new customizations are available from the Component Catalog
    - [x] The component example still has:
        - an accurate preview
        - valid sample code
        - correct keyboard behavior
        - correct and comprehensive guidance around how to use the component
- [x] Changes to the component are tested/the new version of the component is tested
- [x] Component API follows standard patterns in noredink-ui
    - e.g., as a dev, I can conveniently add an `nriDescription`
    - and adding a new feature to the component will _not_ require major API changes to the component

micahhahn commented on src/Nri/Ui/Block/V4.elm:
```diff
@@ -304,6 +329,20 @@ blankWithId =
     BlankWithId
 
 
+{-| Wraps a group of content in a `strong` tag
+-}
+bold : List (Content msg) -> Content msg
+bold =
+    Markdown Bold
+
+
+{-| Wraps a group of content in a `em` tag
+-}
+italic : List (Content msg) -> Content msg
+italic =
+    Markdown Italic
+
+
```
I thought perhaps that `strong` and `emphasis` might be better names semantically, but was worried about `emphasis` causing confusion with how we also us emphasis to describe display elements.

micahhahn commented on src/Content.elm:
```diff
@@ -51,6 +53,22 @@ markdownContent content =
         |> List.map fromUnstyled
 
 
+{-| Provide a string that will be rendered as markdown and placed in a paragraph tag, with the default margin removed.
+-}
+markdownParagraph : String -> List (Html msg)
+markdownParagraph content =
+    case Markdown.Block.parse Nothing content of
+        [ Markdown.Block.Paragraph _ inlines ] ->
+            -- The library always parses into a paragraph and never into `PlainInline`
+            List.map (Markdown.Inline.toHtml >> fromUnstyled) inlines
+                |> p [ css [ Css.margin Css.zero ] ]
+                |> List.singleton
+
+        _ ->
+            Markdown.toHtml Nothing content
+                |> List.map fromUnstyled
+
+
```
I considered simply changing `markdownContent`, but there are enough callers I was scared of breaking other places.

micahhahn commented on src/Nri/Ui/Mark/V2.elm:
I believe I updated all places where we should call `stripMarkdownSyntax`, but please give it a close look to make sure I didn't miss something!

micahhahn commented:
@tesk9 I noticed that `msg` is essentially a phantom type of `Content` now that we have pulled the QB stuff out of `Content`.  I'm guessing we were holding off on removing `msg` until we do a major version bump to `V5`?

tesk9 reviewed (commented):
This is looking great! Thanks for updating the examples and for structuring your PR for a super-straightforward review. 🙏 

---

The only thing I'd like to request is some tests around the "stripMarkdownSyntax" behavior (not the function itself -- just that Block and Highlighter's `::before` and `::after` pseudoelements end up with the right content, and that the sometimes-visually-displayed start tag renders correctly _as markdown_ for highlighter as well.)

Charbel added super useful helpers for testing ::before and ::after content [here](https://github.com/NoRedInk/noredink-ui/blob/master/tests/Spec/Nri/Ui/Highlighter.elm#L352) through line 430. I think pulling these helpers out somewhere for use in the Block specs would be great!

Could you also please add a patch note about the change to Mark for markdown support? Thanks!

---

>  I noticed that msg is essentially a phantom type of Content now that we have pulled the QB stuff out of Content. I'm guessing we were holding off on removing msg until we do a major version bump to V5?

I think this is a mistake that nobody had noticed yet! Or at least, this isn't something that I was keeping track of. I think it's just an oversight. But yeah I think waiting for V5 probably makes sense.

micahhahn commented:
@tesk9 I've added a patch changelog message to Mark and added a couple unit tests - let me know if those aren't comprehensive enough.  (I'm not quite sure I know what you mean by "sometimes-visually-displayed" start tag)

micahhahn commented:
@tesk9 Thank you for the unit test examples! I've merged them in and fixed them along with some other changes we talked about.

tesk9 commented on src/Content.elm:
```diff
@@ -51,6 +53,20 @@ markdownContent content =
 
+{-| Provide a string that will be rendered as markdown and placed in a paragraph tag, with the default margin removed.
+-}
+markdownInline : String -> List (Html msg)
```
Love the rename!

tesk9 commented on src/Content.elm:
```diff
@@ -51,6 +53,20 @@ markdownContent content =
         |> List.map fromUnstyled
 
 
+{-| Provide a string that will be rendered as markdown and placed in a paragraph tag, with the default margin removed.
```
```suggestion
{-| Provide a string that will be rendered as inline markdown, with the default wrapping paragraph removed.
```

tesk9 reviewed (approved):
Thanks! Looks great!
